File: backend/app/utils/filename.py
import re

UNSAFE = re.compile(r'[<>:"/\\|?*\x00-\x1f]')


def sanitize_filename(raw: str | None, fallback: str = "crafted-document") -> str:
    name = (raw or "").strip()
    name = UNSAFE.sub("", name)
    name = name.replace("..", ".")
    name = name.lstrip(".")
    if name.lower().endswith(".pdf"):
        name = name[:-4]
    name = re.sub(r"\s+", " ", name).strip()
    if not name:
        name = fallback
    return f"{name[:180]}.pdf"

File: frontend/src/utils/filename.js
const UNSAFE = /[<>:"/\\|?*\u0000-\u001f]/g

export function sanitizeFilename(raw, fallback = 'crafted-document') {
  let name = String(raw || '').trim()
  name = name.replace(UNSAFE, '')
  name = name.replace(/\.\.+/g, '.')
  name = name.replace(/^\.+/, '')
  name = name.replace(/\.pdf$/i, '')
  name = name.replace(/\s+/g, ' ').trim()
  if (!name) name = fallback
  return `${name.slice(0, 180)}.pdf`
}

export function stripPdfExtension(name) {
  return String(name || '').replace(/\.pdf$/i, '')
}
